agent/php_user_instrument.c

```diff
@@ -114,7 +116,33 @@ int nr_zend_call_orig_execute_special(nruserfn_t* wraprec,
   return zcaught;
 }
 
-#if ZEND_MODULE_API_NO >= ZEND_7_4_X_API_NO
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO
+static inline void nr_php_wraprec_lookup_set(nruserfn_t* wr,
+                                             zend_function* zf) {
+  // for situation when wraprec is added after first execution of the function
+  // store the wraprec in the op_array extension for the duration of the request for later lookup
+  // The op_array extension slot for function may not be initialized yet because it is
+  // initialized only on the first call made to the function in that request. This would
+  // mean that run_time_cache is NULL and wraprec cannot be stored yet! It will be stored
+  // on the first call to the function when observer is registered for that function.
+  zend_init_func_run_time_cache(&zf->op_array);
+  if (NULL != RUN_TIME_CACHE(&zf->op_array)) {
+    ZEND_OP_ARRAY_EXTENSION(&zf->op_array, NR_PHP_PROCESS_GLOBALS(op_array_extension_handle)) = wr;
+  }
+}
+
+static inline nruserfn_t* nr_php_wraprec_lookup_get(zend_function* zf) {
+  nruserfn_t *wraprec = NULL;
+  
+  if (NULL != RUN_TIME_CACHE(&zf->op_array)) {
+    wraprec = ZEND_OP_ARRAY_EXTENSION(&zf->op_array, NR_PHP_PROCESS_GLOBALS(op_array_extension_handle));
+  }
+  if (NULL != wraprec && NRPRG(pid) != wraprec->pid) {
+    wraprec = NULL;
+  }
```

Comment:
I'm not certain this safeguard is needed at all. I need to investigate this a bit more.

Reply:
Doesn't this protect us in case another extension has written to the cache?

Reply:
A method of using `NRPRG(pid)` to validate if memory pointed to by the value retrieved from `ZEND_OP_ARRAY_EXTENSION`, has been allocated by the agent, has been replaced by a unique memory marker in 3cc2311c5278b0d47de0d6f9b461a60a67343d42.